### ISSUE TITLE
mkdir -p for FilePath

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5704]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#5704|
|Type|enhancement|
|Created|2012-06-08 22:23:46Z|
|Branch|https://github.com/twisted/twisted/tree/filepath-mkdir-p-5704-2|

Sometimes I just want to create a directory, you know?  I don't care about EISDIR or any nonsense like that.  Just do it.

Attachments:

* [makedirs.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/45a/45a64e4e14cd7a38839c62b70bff8b8a45c6daa4/5560b6c6319c01800a0dbcb7bac89dd16929a6fe.patch) (582 bytes) - added by mayank on 2013-04-21 22:43:57Z - Possible patch
* [makedirs.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/45a/45a64e4e14cd7a38839c62b70bff8b8a45c6daa4/df65adcaea0ace930be616cf5246fe33302a642e.patch) (1500 bytes) - added by mayank on 2013-04-23 12:49:51Z - Fixed patch
* [5704-joelverhagen.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/45a/45a64e4e14cd7a38839c62b70bff8b8a45c6daa4/b2d42e7a2c6b6f5e30fbd975cc8d7c145e0d7573.patch) (2792 bytes) - added by joelverhagen on 2014-05-29 05:02:26Z - Add mode and exist_ok arguments
* [mkdir-p-filepath-5704-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/45a/45a64e4e14cd7a38839c62b70bff8b8a45c6daa4/2066335f09aa5973fd4c7d0b305fb9a518d0b28e.patch) (3719 bytes) - added by judge_gregg on 2015-12-02 22:09:54Z - 
* [mkdir-p-filepath-5704-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/45a/45a64e4e14cd7a38839c62b70bff8b8a45c6daa4/f510769ee3929ea2f4e89521315188ae357242c3.patch) (3396 bytes) - added by judge_gregg on 2015-12-03 20:07:37Z - 
* [mkdir-p-filepath-5704-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/45a/45a64e4e14cd7a38839c62b70bff8b8a45c6daa4/9f61af249dc77e99bb39521b696842f7a29d4454.patch) (1543 bytes) - added by judge_gregg on 2015-12-07 20:34:00Z - 
* [mkdir-p-filepath-5704-4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/45a/45a64e4e14cd7a38839c62b70bff8b8a45c6daa4/308ebf8dcee39b1efcf54d9542d5397b0f2822eb.patch) (5156 bytes) - added by judge_gregg on 2015-12-13 16:24:45Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5704 5704
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_filepath_mkdir_p_5704_2 branches/filepath-mkdir-p-5704-2
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1339194226000000 1339194226000000
changetime__1450104156586301 1450104156586301
version__None None
owner__judge_gregg judge_gregg
cc__hynek
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jinrummie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jinrummie set owner to jinrummie<br>jinrummie set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="mayank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|mayank commented|
|-|-|

Doesn't twisted.python.filepath.FilePath.makedirs already do this?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

It does not.  It raises an exception if the directory already exists.

**Comment by twisted-trac** [comment]
|<img alt="mayank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|mayank set owner to mayank<br>mayank set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_7"></a>|@Julian commented|
|-|-|

Hey!

Thanks for working on this.

I think you're in the right place, but for backwards compatibility reasons (and also because though I agree that this is useful, it doesn't need to be the default behavior) we can't just change that method to do this.

However, if you take a gander at the Python 3 documentation (don't worry, no one will yell), you'll find that `os.makedirs` now has an `exist_ok` flag which does this.http://docs.python.org/3/library/os.html?highlight=makedirs#os.makedirs

I think the reasonable thing to do here is to add the same argument to `FilePath.makedirs`. It won't be enough to just pass that argument along, since that won't work for Python2, so I'd recommend ignoring the argument and just re-implementing it using `errno.EEXIST` and a `try/except`. So, in short, have the same interface, with an arg called `exist_ok`, but in a way that works in Py2 as well.

I should note that now might be a good time to add the `mode` argument (which is present even on Py2). Especially so since adding arguments to makedirs will cause backwards compatibility issues later if we want to reorder them. (Oh and, a minor point, if you remember, when you put the patch up for review, click reassign and reassign it to the empty name at the top of the list, so that it's clear that no one is working on it at the moment :D)

Also, most importantly, your patch should include tests to test the code you're adding. See http://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted if you haven't already.

Let us know if you have any questions.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_8"></a>|@Julian commented|
|-|-|

Oh -- one more specific note, I see your implementation is more or less what will happen, though now only with a flag argument, but you don't need to create a new exception object. You can reraise an existing exception by using the raise statement without any "arguments". E.g., it should look like:

```
except OSError as error:
    if error.errno != errno.EEXIST and error.errno != errno.EISDIR:
        raise
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

A couple more thoughts:

  1. If an option is added, it shouldn't be called `exists_ok` since that doesn't follow the Twisted naming convention.
  1. I think it's worth some particular consideration whether adding an option satisfies the intent of the ticket.  If this is something that has to be explicitly passed in, then it still has to be considered.  `FilePath.makedirs()` is still an attractive nuisance.  It'll be easy to write code that fails when a directory exists already, since the default is still to fail in that case.  Perhaps that's just what we'll have to do, given that `FilePath.makedirs` exists already with some behavior, but please think about adding a new method with the new behavior instead, and perhaps finding some way to encourage its use instead of `makedirs`.  (I think it's fine if this idea is rejected, but it should be explicitly rejected rather than ignored)

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_10"></a>|@Julian commented|
|-|-|

For 1, doing

```
def makedirs(self, **kwargs):
    try:
        os.makedirs(self.path, **kwargs)
    except OSError as error:
        if not kwargs.get("exists_ok", False) or ...:
            raise
```

would be something like the implementation I was suggesting, along with documenting that this function wraps `os.makedirs`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

Thanks for clarifying, Julian.  I don't think that's an API direction we want to go very far in (if we want to go there at all).  For one thing, there are multiple implementations of (at least parts of) the `FilePath` API - and ideally at some point we'll make the interface explicit.  Having a `**kwargs`, the meaning of which is interpreted by `os.makedirs`, is not very compatible with either of these ideas.

**Comment by twisted-trac** [comment]
|<img alt="mayank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|mayank removed owner|
|-|-|

I am not sure about the nomenclature but here is a patch with unittest.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_13"></a>|@hynek commented|
|-|-|

Hi Mayank,

thank you for your contribution!

In order for your patch to be accepted, it will have to follow the contribution guidelines outlined in http://twistedmatrix.com/trac/wiki/ReviewProcess which your patch currently violates a few of:

 1. Your patch creates a new test file. Tests for FilePath go into `twisted/test/test_paths`. Pro tip: the header of `twisted/python/filepath.py` gives you a hint here: `# -*- test-case-name: twisted.test.test_paths -*-`. I use these headers as pointers where to look.
 1. The tests have no docstrings.
 1. The tests don’t verify anything. If your `makedirsForced` were a NOP, they would still pass. 
 1. Why does the patch remove the `return` statement from `makedirs`?
 1. `makedirsForced` has no docstring.
 1. I find the name rather questionable, there’s nothing forced about it. It’s just ignoring errors. So `makedirsLiberal` would be more fitting. ;)
 1. `os.makedirs()` has no argument `exists`. As of Python 3.2, it has an `exists_ok` that would save all the catching. Is this the reason for you catching (always) `TypeError`?


Thanks again!


----

I feel like this ticket needs to settle whether to use a method or an argument and maybe how to call it.


Personally, I would abandon that makedirs nonsense (who cares how it is implemented?) and concentrate on `createDirectory`. Why not add options `recursive` (maybe something better?) and `existsOk`? Since the method currently doesn’t take any arguments, it seems safe to do so.

**Comment by twisted-trac** [comment]
|<img alt="joelverhagen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|joelverhagen commented|
|-|-|

I gave a hack at this enhancement. Let me know if anything should be changed.

Replying to [Julian](#note_7):
> I should note that now might be a good time to add the mode argument (which is present even on Py2).

Great idea. Done.

Replying to [exarkun](#note_9):
> A couple more thoughts:
> 
>   1. If an option is added, it shouldn't be called `exists_ok` since that doesn't follow the Twisted naming convention.

I opted to leave the keyword argument matching Python standard library, since the function name itself seems to be doing that. I am by no means set on this decision.

Replying to [hynek](#note_13):
> Why does the patch remove the return statement from makedirs? 

The standard library code for `makedirs` has no return statement, and thus returns `None`.

> Personally, I would abandon that makedirs nonsense (who cares how it is implemented?) and concentrate on createDirectory. Why not add options recursive (maybe something better?) and existsOk? Since the method currently doesn’t take any arguments, it seems safe to do so. 

I can do this route as well, but this seems to be a much deeper API change. If I were to do this, how would I mark `makedirs` as deprecated?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph set owner to joelverhagen|
|-|-|

Replying to [joelverhagen](#note_15):
> I gave a hack at this enhancement. Let me know if anything should be changed.

Thanks for your contribution.  I apologize for the latency you encountered on this review.  We are making a concerted effort to reduce that in the future :-).

> Replying to [Julian](#note_7):
> > I should note that now might be a good time to add the mode argument (which is present even on Py2).
> 
> Great idea. Done.
> 
> Replying to [exarkun](#note_9):
> > A couple more thoughts:
> > 
> >   1. If an option is added, it shouldn't be called `exists_ok` since that doesn't follow the Twisted naming convention.
> 
> I opted to leave the keyword argument matching Python standard library, since the function name itself seems to be doing that. I am by no means set on this decision.

I'm sorry that this was not communicated clearly: a parameter called "`exists_ok`" does not follow the Twisted naming convention, therefore it does not comply with the coding standard and therefore it is not permitted in a new Twisted API.  exarkun was making a statement of policy, not giving you a suggestion to take under advisement :).  You could of course argue that the policy could be changed, but that would be a broader discussion for the mailing list, not something to be decided in the scope of a ticket comment.

> Replying to [hynek](#note_13):
> > Why does the patch remove the return statement from makedirs? 
> 
> The standard library code for `makedirs` has no return statement, and thus returns `None`.

Again, the behavior of the standard library is not really a guiding principle in how Twisted APIs should be designed.  (Many things in Twisted - FilePath included - exist because the code quality in the standard library is highly variable and mostly bad.)  If the code previously returned a value, then it must continue returning a value to adhere with the CompatibilityPolicy.

> > Personally, I would abandon that makedirs nonsense (who cares how it is implemented?) and concentrate on createDirectory. Why not add options recursive (maybe something better?) and existsOk? Since the method currently doesn’t take any arguments, it seems safe to do so. 
> 
> I can do this route as well, but this seems to be a much deeper API change. If I were to do this, how would I mark `makedirs` as deprecated?

There are various utilities in `twisted.python.deprecate` to make this relatively easy.  Luckily, `IFilePath` doesn't have a `makedirs` attribute, so you don't need to face the stickier issue of deprecating part of an interface rather than a class :).

Once again, I appreciate your contribution and I'm sorry we couldn't review it more promptly before.  I hope that you'll resubmit and we can move this forward!

**Comment by twisted-trac** [comment]
|<img alt="judge_gregg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|judge_gregg removed owner|
|-|-|

This patch adds a few cosmetic changes to joelverhagen's patch (mostly compatibility policy and improved docstrings). Let me know if I should add or update anything else.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_19"></a>|@adiroiban commented|
|-|-|

(In [46383]) Branching to filepath-mkdir-p-5704.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_20"></a>|@adiroiban set owner to judge_gregg|
|-|-|

Many thanks for your patch!

Please note that the topfile news file fragment will be concatenated into a single. You should arrange the content of that file so that it can fit a single NEWS / release notes item.

----

It looks like this patch is doing more than the ticket description as it also adds the the `mode` argument.

Either update the ticket description to match the scope of the patch or break it into a separate tickets.

I prefer to have it in a separate ticket. 

For me, the mode Unix flag is "controversial" as it is not OS agnostic and on Windows the code might behave differently. 

Regarding the mode, please also check twisted.python.filepath.Permissions and twisted.python.filepath.RWX


-------

Is `existsOk` the right name ? ... maybe existsOK

------

I think that the code should be written based on - EAFP: Easier to ask for forgiveness than permission.

That is, just try to create it and ignore the exception.

With this in mind, maybe the flag should be ignoreAlreadyExist ... or something like that.

------

This is just a side note.

I would argue that flag arguments are bad design, the stdlib method name is bad and we should have a FilePath.createDirectories() which will always ignore the EISDIR error.

I don't see why you would want to call createDirectories and have an error raised if they already exists.

We can also have createDirectory which will not do recursive creation and might fail if path already exists.

-----

I have sent your patch to our buildbot.

Please check my comments and resubmit a new patch.

Many thanks for your contribution.

**Comment by twisted-trac** [comment]
|<img alt="judge_gregg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|judge_gregg commented|
|-|-|

Replying to [adiroiban](#note_20):

Thank you for your review. I added a new patch taking your comments into account.
----
> We can also have createDirectory which will not do recursive creation and might fail if path already exists.
These 2 options (makedirs vs. createDirectory) seem to have been in discussion since the ticket was opened. I opted for the first, mostly because it seemed easier to me, but I can also give a try to the createDirectory option. I guess if we choose to focus on createDirectory, then there is no need to create the 'mode' ticket as you suggested above ?

**Comment by twisted-trac** [comment]
|<img alt="judge_gregg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|judge_gregg removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_23"></a>|@adiroiban commented|
|-|-|

(In [46389]) Branching to filepath-mkdir-p-5704-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_24"></a>|@adiroiban set owner to judge_gregg|
|-|-|

Thanks for latest patch.

We already have createDirectory() so for me it make sense to also have createDirectories() and not have this (arguably ugly) ignoreAlreadyExist flag :)

-------

The mode flag is a different discussion.

--------

Should this be ignoreAlreadyExist or ignoreAlreadyExists ?

There is no tests to check that non EEXIST OSErrors are raised when ignoreAlreadyExist is True.

Also I wrote a simple tests for calling mkdir for a patch which is already a file. I think that ignoreAlreadyExist should only ignore if path is a dir not a file. The new test fails.

self.assertRaises will return a context manager which had the 'exception' member... not directly the exception. Your assertion is never called as makedirs throws an exception and remaining code from the block is not called.

Assertion should be called outside of the block.... and you should write your test first and see them fail and only later update the code :)

I have converted your self.assertRaises context manager into a much simpler assertion supported only by trial.

------

As a minor comment. There is not much value to have 'check that' , 'test that' in the test's docstring. See https://jml.io/pages/test-docstrings.html

-----

I have updated the test names to be in sync with Twisted naming convention.

Please check my comments and the my latest commits and write a test in which OSErrors are raised if they are not EEXIST 

If possible please submit a patch on top of filepath-mkdir-p-5704-2.

Many thanks!

**Comment by twisted-trac** [comment]
|<img alt="judge_gregg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_26"></a>|judge_gregg removed owner|
|-|-|

Thank you for you review. I added a patch taking your comments into account, on top of filepath-mkdir-5704-2.

The test with a path as file now passes. I also added a new test for OSErrors which are not EEXIST.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_27"></a>|@adiroiban commented|
|-|-|

Changes looks good. Thanks!

I have made a small change in which self.isdir() is called last in the condition. I hope that this can reduce the need for another OS call when ignoreAlreadyExist is False.

Should the flag be named ignoreAlreadyExist or ignoreAlreadyExists ?

Leaving this for review to get the feedback from on more people... maybe some of the people which have already reviewed the previous patches :)

Many thanks for your contribution!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_28"></a>|@hawkowl set owner to judge_gregg|
|-|-|

Hi!

Thanks for the work, everyone. A few things:

- I don't like the flag name. Something like "ignoreExistingDirectories" might work nicer, as it says what we are actually ignoring.

- https://github.com/twisted/twisted/compare/trunk...filepath-mkdir-p-5704-2#diff-c7ada4588c4ea33358407df5c6f0d592R1427 looks a bit strange and hard to understand to me. I think it's the final ```):``` being on a new line.

- https://github.com/twisted/twisted/compare/trunk...filepath-mkdir-p-5704-2#diff-9a142283adadbf94b8c76a8b01120012R1411 This shouldn't have a b prefix, although it doesn't actually matter in this situation.

Please fix these points and submit for review, so a core dev can double check and merge it. Please base your changes off the existing branch if possible :)

**Comment by twisted-trac** [comment]
|<img alt="judge_gregg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_29"></a>|judge_gregg removed owner|
|-|-|

Thank you for your review :).

Patch 4 should fix the points you mentionned above. It is based off the
existing branch.

With regards to the flag name, the flag is needed only if the last directory in
the path already exists. The previous directories are ignored by makedirs, even
without the flag.

So maybe ignoreExistingDirectory would be more appropriate ?

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_30"></a>|@adiroiban commented|
|-|-|

Thanks for the patch :)

+1 for ignoreExistingDirectory

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_31"></a>|@adiroiban set owner to judge_gregg|
|-|-|

I have applied your patch, I have also applied the ignoreExistingDirectory change.

I have sent it to our buildbot builders.

If all is ok I will merge. Many thanks for your contribution.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_32"></a>|@adiroiban commented|
|-|-|

Tests are green.

judge_gregg can you please check my latest changes and make sure you are happy with them?

If all is ok I will merge.

Many thanks!

**Comment by twisted-trac** [comment]
|<img alt="judge_gregg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_33"></a>|judge_gregg commented|
|-|-|

Looks good to me.
Thanks a lot for your help :).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_34"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46503]) Merge filepath-mkdir-p-5704-2: Handle LoopingCall.withCount with interval 0.

Author: judge_gregg
Reviewer: adiroiban, hawkowl
Fixes: [#5704](https://github.com/twisted/twisted/issues/5704)
